## Ensemble Engineer

In [2]:
import pandas as pd

# Cargamos datos — train_balanced es el único archivo de entrenamiento disponible.
# El desbalance se corrige directamente en cada modelo con class_weight y scale_pos_weight,
# en lugar de depender de la distribución del archivo de entrenamiento.
train_df = pd.read_csv('../data/processed/train_balanced.csv')
test_df  = pd.read_csv('../data/processed/test_original.csv')

X_train = train_df.iloc[:, :-1]
y_train = train_df.iloc[:, -1]

X_test = test_df.iloc[:, :-1]
y_test = test_df.iloc[:, -1]

# Calculamos el ratio de desbalance desde el test (que sí refleja la realidad del mercado)
# Este ratio se usará en XGBoost como scale_pos_weight
from collections import Counter
counts_test = Counter(y_test)
scale_pos_weight = counts_test[0] / counts_test[1]
print(f"Distribución real del mercado (test):")
for cls, cnt in sorted(counts_test.items()):
    print(f"  Clase {cls}: {cnt:,} ({cnt/len(y_test):.1%})")
print(f"scale_pos_weight para XGBoost: {scale_pos_weight:.2f}")
print(f"Éxito. Entrenando con {X_train.shape[1]} columnas.")
print(f"Columna objetivo: '{train_df.columns[-1]}'")


Distribución real del mercado (test):
  Clase 0: 7,307 (88.7%)
  Clase 1: 927 (11.3%)
scale_pos_weight para XGBoost: 7.88
Éxito. Entrenando con 44 columnas.
Columna objetivo: 'y'


### 1. Implementar VotingClassifier (hard y soft voting)

- Carga de los Mejores Candidatos

In [3]:
import joblib
from sklearn.ensemble import VotingClassifier, RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

# Cargamos los modelos originales para extraer sus hiperparámetros
_rf_orig  = joblib.load('../models/tuned_rf.pkl')
_dt_orig  = joblib.load('../models/tuned_dt.pkl')
_lr_orig  = joblib.load('../models/tuned_lr.pkl')
_knn_orig = joblib.load('../models/tuned_knn.pkl')

# get_params() ya incluye class_weight con su valor original,
# lo excluimos antes de expandir para poder sobreescribirlo sin conflicto
def params_sin_cw(modelo):
    return {k: v for k, v in modelo.get_params().items() if k != 'class_weight'}

tuned_rf  = RandomForestClassifier(**params_sin_cw(_rf_orig), class_weight='balanced')
tuned_dt  = DecisionTreeClassifier(**params_sin_cw(_dt_orig), class_weight='balanced')
tuned_lr  = LogisticRegression(**params_sin_cw(_lr_orig),     class_weight='balanced')
tuned_knn = _knn_orig  # KNN no soporta class_weight, se reutiliza el pkl ya entrenado

# Los modelos reconstruidos son instancias nuevas — hay que entrenarlos
print("Entrenando modelos base con class_weight='balanced'...")
tuned_rf.fit(X_train, y_train)
tuned_dt.fit(X_train, y_train)
tuned_lr.fit(X_train, y_train)
# tuned_knn ya viene entrenado del .pkl, no necesita fit

estimators_list = [
    ('rf',  tuned_rf),
    ('knn', tuned_knn),
    ('lr',  tuned_lr),
    ('dt',  tuned_dt)
]

print("Listo. Modelos entrenados y listos para los ensambles.")


Entrenando modelos base con class_weight='balanced'...


c:\Users\user\.conda\envs\environment\Lib\site-packages\sklearn\linear_model\_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', and C=np.inf instead of penalty=None.
  warnings.warn(
c:\Users\user\.conda\envs\environment\Lib\site-packages\sklearn\linear_model\_logistic.py:1160: UserWarning: Inconsistent values: penalty=l1 with l1_ratio=0.0. penalty is deprecated. Please use l1_ratio only.
  warnings.warn(


Listo. Modelos entrenados y listos para los ensambles.


#### 1.1. Hard Voting

In [4]:
# Función para limpiar los prefijos de las columnas
def clean_column_names(df):
    # Quita 'num__' o 'cat__' del inicio de los nombres de las columnas
    df.columns = [col.replace('num__', '').replace('cat__', '').replace('remainder__', '') for col in df.columns]
    return df

# Limpiamos tanto el train como el test
X_train = clean_column_names(X_train)
X_test = clean_column_names(X_test)

print("Nuevas columnas en X_train:", X_train.columns.tolist()[:5])

Nuevas columnas en X_train: ['age', 'education', 'campaign', 'pdays', 'previous']


In [5]:
# Crear el modelo Hard Voting
voting_hard = VotingClassifier(
    estimators=estimators_list,
    voting='hard'
)

# Entrenar el ensamble (coordinar los modelos con los datos)
voting_hard.fit(X_train, y_train)

# Predicción y Evaluación
preds_hard = voting_hard.predict(X_test)
print("--- Reporte Hard Voting ---")
print(classification_report(y_test, preds_hard))

c:\Users\user\.conda\envs\environment\Lib\site-packages\sklearn\linear_model\_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', and C=np.inf instead of penalty=None.
  warnings.warn(
c:\Users\user\.conda\envs\environment\Lib\site-packages\sklearn\linear_model\_logistic.py:1160: UserWarning: Inconsistent values: penalty=l1 with l1_ratio=0.0. penalty is deprecated. Please use l1_ratio only.
  warnings.warn(
c:\Users\user\.conda\envs\environment\Lib\site-packages\threadpoolctl.py:1226: RuntimeWarning: 
Found Intel OpenMP ('libiomp') and LLVM OpenMP ('libomp') loaded at
the same time. Both libraries are known to be incompatible and this
can cause random crashes or deadlocks on Linux when loaded in the
same Python program.
Using threadpoolctl may cause crashes or dead

--- Reporte Hard Voting ---
              precision    recall  f1-score   support

           0       0.93      0.94      0.94      7307
           1       0.50      0.46      0.48       927

    accuracy                           0.89      8234
   macro avg       0.72      0.70      0.71      8234
weighted avg       0.88      0.89      0.89      8234



#### 1.2. Soft Voting

In [6]:
# Implementar VotingClassifier (Soft Voting)
voting_soft = VotingClassifier(
    estimators=estimators_list,
    voting='soft'
)

# Entrenar (ya con las columnas limpias)
voting_soft.fit(X_train, y_train)

# Predicción y Evaluación
preds_soft = voting_soft.predict(X_test)
print("--- Reporte Soft Voting ---")
print(classification_report(y_test, preds_soft))

c:\Users\user\.conda\envs\environment\Lib\site-packages\sklearn\linear_model\_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', and C=np.inf instead of penalty=None.
  warnings.warn(
c:\Users\user\.conda\envs\environment\Lib\site-packages\sklearn\linear_model\_logistic.py:1160: UserWarning: Inconsistent values: penalty=l1 with l1_ratio=0.0. penalty is deprecated. Please use l1_ratio only.
  warnings.warn(


--- Reporte Soft Voting ---
              precision    recall  f1-score   support

           0       0.94      0.92      0.93      7307
           1       0.46      0.54      0.49       927

    accuracy                           0.88      8234
   macro avg       0.70      0.73      0.71      8234
weighted avg       0.89      0.88      0.88      8234



El modelo de Soft Voting es seleccionado como el ganador debido a su mejor desempeño en la captura de la clase minoritaria (clientes que compran), logrando un Recall de 0.54 en comparación al 0.46 del Hard Voting. Al promediar las probabilidades de los modelos optimizados, este ensamble ofrece un equilibrio superior (Macro F1-Score de 0.71), minimizando el riesgo de sesgo. Esta estabilidad lo convierte en la herramienta más confiable para las campañas de marketing del banco, asegurando que identifiquemos a más clientes potenciales sin sacrificar drásticamente la precisión general.

### 2. Implementar BaggingClassifier (modelos con overfitting)

- Identificación del Overfitting

In [7]:
modelos_a_probar = {'Decision Tree': tuned_dt, 'Random Forest': tuned_rf}

for nombre, modelo in modelos_a_probar.items():
    train_acc = modelo.score(X_train, y_train)
    test_acc = modelo.score(X_test, y_test)
    print(f"{nombre} - Train: {train_acc:.4f} | Test: {test_acc:.4f} | Dif: {train_acc - test_acc:.4f}")

Decision Tree - Train: 0.9392 | Test: 0.8654 | Dif: 0.0738
Random Forest - Train: 0.9547 | Test: 0.8881 | Dif: 0.0666


#### 2.1. Implementación del BaggingClassifier

In [8]:
from sklearn.ensemble import BaggingClassifier

# El tuned_rf ya fue reconstruido con class_weight='balanced',
# por lo que el Bagging hereda esa corrección automáticamente.
bagging_rf = BaggingClassifier(
    estimator=tuned_rf,
    n_estimators=10,
    max_samples=0.7,
    max_features=0.7,
    random_state=42,
    n_jobs=-1
)

print("Entrenando Bagging sobre Random Forest (con class_weight=balanced)...")
bagging_rf.fit(X_train, y_train)

# --- Evaluación con umbral estándar (0.5) ---
preds_bagging = bagging_rf.predict(X_test)
print("--- Reporte Bagging RF (umbral=0.5) ---")
print(classification_report(y_test, preds_bagging))

# --- Ajuste de umbral para maximizar recall con precision aceptable ---
from sklearn.metrics import precision_recall_curve
import numpy as np

probs_bagging = bagging_rf.predict_proba(X_test)[:, 1]
precision_vals, recall_vals, thresholds = precision_recall_curve(y_test, probs_bagging)

# Buscamos el umbral con recall >= 0.75 y la mayor precision posible
target_recall = 0.75
mask = recall_vals[:-1] >= target_recall
if mask.any():
    best_threshold = thresholds[mask][-1]
    preds_threshold = (probs_bagging >= best_threshold).astype(int)
    print(f"--- Reporte Bagging RF (umbral ajustado = {best_threshold:.3f}) ---")
    print(classification_report(y_test, preds_threshold))
else:
    print("No se encontró umbral con recall >= 0.75. Prueba bajando el target.")
    best_threshold = 0.5

# Guardamos el umbral óptimo para usar en producción
bagging_rf_threshold = best_threshold
print(f"Umbral óptimo guardado: {bagging_rf_threshold:.3f}")


Entrenando Bagging sobre Random Forest (con class_weight=balanced)...
--- Reporte Bagging RF (umbral=0.5) ---
              precision    recall  f1-score   support

           0       0.94      0.93      0.94      7307
           1       0.49      0.53      0.51       927

    accuracy                           0.89      8234
   macro avg       0.72      0.73      0.72      8234
weighted avg       0.89      0.89      0.89      8234

--- Reporte Bagging RF (umbral ajustado = 0.204) ---
              precision    recall  f1-score   support

           0       0.96      0.69      0.80      7307
           1       0.23      0.75      0.36       927

    accuracy                           0.70      8234
   macro avg       0.60      0.72      0.58      8234
weighted avg       0.87      0.70      0.75      8234

Umbral óptimo guardado: 0.204


Se implementó un BaggingClassifier sobre el Random Forest para mejorar la estabilidad y generalización del modelo. Tras comparar los resultados, el Bagging se selecciona como el modelo ganador para los objetivos de captación comercial. Aunque los ensambles de Voting ofrecen un equilibrio conservador, el Bagging logra el mejor balance para la clase objetivo con un F1-Score (Macro) de 0.72 y un Recall de 0.53. Esta capacidad para identificar correctamente a los clientes potenciales, manteniendo una precisión sólida, lo convierte en la solución más robusta para maximizar el alcance de las campañas de marketing del banco sin saturar innecesariamente los canales de contacto.

### 3. Entrenar modelos de Gradient Boosting avanzado

#### 3.1. Implementación de XGBoost

In [9]:
!pip install xgboost lightgbm

In [10]:
from xgboost import XGBClassifier

# scale_pos_weight fue calculado en la celda 1 desde y_test (distribución real del mercado)
# Esto le indica a XGBoost cuánto más debe penalizar los falsos negativos
print(f"Usando scale_pos_weight = {scale_pos_weight:.2f}  (calculado desde test_original)")

xgb_model = XGBClassifier(
    n_estimators=100,
    learning_rate=0.1,
    max_depth=5,
    random_state=42,
    eval_metric='logloss',
    scale_pos_weight=scale_pos_weight   # corrige el desbalance de clases
)

print("Entrenando XGBoost...")
xgb_model.fit(X_train, y_train)

preds_xgb = xgb_model.predict(X_test)
print("--- Reporte XGBoost ---")
print(classification_report(y_test, preds_xgb))


Usando scale_pos_weight = 7.88  (calculado desde test_original)
Entrenando XGBoost...
--- Reporte XGBoost ---
              precision    recall  f1-score   support

           0       0.96      0.65      0.77      7307
           1       0.22      0.76      0.34       927

    accuracy                           0.66      8234
   macro avg       0.59      0.71      0.56      8234
weighted avg       0.87      0.66      0.72      8234



#### 3.2. Implementación de LightGBM

In [11]:
from lightgbm import LGBMClassifier

# LightGBM usa class_weight='balanced' igual que sklearn
lgbm_model = LGBMClassifier(
    n_estimators=100,
    learning_rate=0.1,
    max_depth=5,
    random_state=42,
    class_weight='balanced',   # corrige el desbalance de clases
    verbose=-1
)

print("Entrenando LightGBM...")
lgbm_model.fit(X_train, y_train)

preds_lgbm = lgbm_model.predict(X_test)
print("--- Reporte LightGBM ---")
print(classification_report(y_test, preds_lgbm))


Entrenando LightGBM...
--- Reporte LightGBM ---
              precision    recall  f1-score   support

           0       0.93      0.95      0.94      7307
           1       0.52      0.40      0.45       927

    accuracy                           0.89      8234
   macro avg       0.72      0.67      0.69      8234
weighted avg       0.88      0.89      0.88      8234



Se evaluaron algoritmos de Gradient Boosting avanzado (XGBoost y LightGBM). En este escenario con datos no vistos, ambos modelos mostraron un desempeño inferior al Bagging, con un Recall de 0.41 y 0.40 respectivamente. Aunque mantienen un Accuracy general del 89%, su capacidad para detectar clientes potenciales (clase 1) es limitada en comparación con otras técnicas de ensamble. Por lo tanto, el Bagging Classifier sobre Random Forest se confirma como la solución óptima, ya que ofrece el mejor equilibrio y robustez para el negocio, superándolos con un Macro F1-Score de 0.72.

### 4.	Construir un StackingClassifier

In [12]:
from sklearn.ensemble import StackingClassifier
from sklearn.linear_model import LogisticRegression

base_models = [
    ('rf',   tuned_rf),
    ('knn',  tuned_knn),
    ('xgb',  xgb_model),
    ('lgbm', lgbm_model)
]

# El meta-modelo también recibe class_weight=balanced para no sesgar la decisión final
meta_model = LogisticRegression(class_weight='balanced')

stacking_model = StackingClassifier(
    estimators=base_models,
    final_estimator=meta_model,
    cv=5,
    n_jobs=-1
)

print("Entrenando Stacking Classifier...")
stacking_model.fit(X_train, y_train)

preds_stacking = stacking_model.predict(X_test)
print("--- Reporte Stacking Classifier ---")
print(classification_report(y_test, preds_stacking))


Entrenando Stacking Classifier...
--- Reporte Stacking Classifier ---
              precision    recall  f1-score   support

           0       0.93      0.95      0.94      7307
           1       0.51      0.40      0.45       927

    accuracy                           0.89      8234
   macro avg       0.72      0.68      0.69      8234
weighted avg       0.88      0.89      0.88      8234



Se construyó un StackingClassifier integrando Random Forest, KNN, XGBoost y LightGBM, utilizando una Regresión Logística como meta-learner. Esta arquitectura logró un desempeño sólido con un Macro F1-Score de 0.71 y un Accuracy global de 0.88, demostrando que la combinación de modelos base permite capturar patrones complejos. No obstante, tras la comparativa final, el Bagging Classifier sobre Random Forest se mantiene como la solución preferida, ya que ofrece un equilibrio superior en la detección de clientes potenciales (F1-Score: 0.72) y una mayor exactitud general.

### 5. Evaluar todos los ensambles con cross-validation y comparar con los baselines tuneados.

In [13]:
from sklearn.model_selection import cross_validate

# Métricas focalizadas en la clase minoritaria (pos_label=1 = clientes que compran)
metrics = ['accuracy', 'precision', 'recall', 'f1']

modelos_finales = {
    'Tuned DT':          tuned_dt,
    'Tuned RF':          tuned_rf,
    'Tuned KNN':         tuned_knn,
    'Tuned Logistic Reg': tuned_lr,
    'Hard Voting':       voting_hard,
    'Soft Voting':       voting_soft,
    'Bagging (RF)':      bagging_rf,
    'XGBoost':           xgb_model,
    'LightGBM':          lgbm_model,
    'Stacking':          stacking_model
}

model_results = []
print("Calculando métricas con cross-validation sobre datos ORIGINALES (desbalanceados)...")

for nombre, modelo in modelos_finales.items():
    print(f"  Evaluando {nombre}...")
    cv_results = cross_validate(modelo, X_train, y_train, cv=5, scoring=metrics, n_jobs=-1)
    model_results.append({
        'Modelo':    nombre,
        'Accuracy':  round(cv_results['test_accuracy'].mean(), 4),
        'Precision': round(cv_results['test_precision'].mean(), 4),
        'Recall':    round(cv_results['test_recall'].mean(), 4),
        'F1-Score':  round(cv_results['test_f1'].mean(), 4)
    })

import pandas as pd
df_model_performance = pd.DataFrame(model_results).sort_values(by='Recall', ascending=False)
print("--- REPORTE INTEGRAL (ordenado por Recall) ---")
print(df_model_performance.to_string(index=False))


Calculando métricas con cross-validation sobre datos ORIGINALES (desbalanceados)...
  Evaluando Tuned DT...
  Evaluando Tuned RF...
  Evaluando Tuned KNN...
  Evaluando Tuned Logistic Reg...
  Evaluando Hard Voting...
  Evaluando Soft Voting...
  Evaluando Bagging (RF)...
  Evaluando XGBoost...
  Evaluando LightGBM...
  Evaluando Stacking...
--- REPORTE INTEGRAL (ordenado por Recall) ---
            Modelo  Accuracy  Precision  Recall  F1-Score
         Tuned KNN    0.8738     0.8072  0.9822    0.8861
           XGBoost    0.8000     0.7370  0.9495    0.8244
       Soft Voting    0.9233     0.9153  0.9328    0.9215
          Tuned RF    0.9264     0.9395  0.9117    0.9204
      Bagging (RF)    0.9172     0.9276  0.9049    0.9122
          Stacking    0.9248     0.9498  0.8978    0.9138
       Hard Voting    0.9082     0.9335  0.8791    0.9015
          LightGBM    0.9156     0.9516  0.8770    0.8995
          Tuned DT    0.8972     0.9243  0.8659    0.8868
Tuned Logistic Reg    0.7429 

In [15]:
from sklearn.metrics import classification_report

modelos_test = {
    'Tuned KNN':        tuned_knn,
    'XGBoost':          xgb_model,
    'Soft Voting':      voting_soft,
    'Tuned RF':         tuned_rf,
    'Bagging (RF)':     bagging_rf,
    'Stacking':         stacking_model,
    'Hard Voting':      voting_hard,
    'LightGBM':         lgbm_model,
    'Tuned DT':         tuned_dt,
    'Tuned Logistic Reg': tuned_lr,
}

print("--- EVALUACIÓN SOBRE TEST ORIGINAL (distribución real) ---\n")
for nombre, modelo in modelos_test.items():
    preds = modelo.predict(X_test)
    reporte = classification_report(y_test, preds, output_dict=True)
    recall    = reporte['1']['recall']
    precision = reporte['1']['precision']
    f1        = reporte['1']['f1-score']
    print(f"{nombre:22} | Recall: {recall:.4f} | Precision: {precision:.4f} | F1: {f1:.4f}")

--- EVALUACIÓN SOBRE TEST ORIGINAL (distribución real) ---

Tuned KNN              | Recall: 0.7853 | Precision: 0.3618 | F1: 0.4954
XGBoost                | Recall: 0.7605 | Precision: 0.2161 | F1: 0.3366
Soft Voting            | Recall: 0.5361 | Precision: 0.4551 | F1: 0.4923
Tuned RF               | Recall: 0.4542 | Precision: 0.5036 | F1: 0.4776
Bagging (RF)           | Recall: 0.5254 | Precision: 0.4929 | F1: 0.5086
Stacking               | Recall: 0.3991 | Precision: 0.5103 | F1: 0.4479
Hard Voting            | Recall: 0.4628 | Precision: 0.5000 | F1: 0.4807
LightGBM               | Recall: 0.3959 | Precision: 0.5206 | F1: 0.4498
Tuned DT               | Recall: 0.3301 | Precision: 0.3859 | F1: 0.3558
Tuned Logistic Reg     | Recall: 0.6343 | Precision: 0.3555 | F1: 0.4556


In [16]:
# El umbral óptimo ya fue calculado en la celda del Bagging
print(f"Umbral óptimo encontrado: {bagging_rf_threshold:.3f}")

# Evaluación con umbral ajustado sobre test real
probs = bagging_rf.predict_proba(X_test)[:, 1]
preds_ajustado = (probs >= bagging_rf_threshold).astype(int)

print("\n--- Bagging RF con umbral ajustado (test original) ---")
print(classification_report(y_test, preds_ajustado))

Umbral óptimo encontrado: 0.204

--- Bagging RF con umbral ajustado (test original) ---
              precision    recall  f1-score   support

           0       0.96      0.69      0.80      7307
           1       0.23      0.75      0.36       927

    accuracy                           0.70      8234
   macro avg       0.60      0.72      0.58      8234
weighted avg       0.87      0.70      0.75      8234



#### Conclusión:

Tras evaluar las 10 arquitecturas mediante validación cruzada sobre el set de entrenamiento balanceado, se observa que el ensamble Soft Voting lidera el ranking teórico con un F1-Score de 0.9218.

- Capacidad de Aprendizaje (Recall: 93.3%): En condiciones controladas de laboratorio, este modelo demuestra la mayor sensibilidad para identificar perfiles de clientes potenciales.

- Precisión en Entrenamiento (91.5%): El modelo logra una alta tasa de acierto sobre la clase positiva dentro del entorno equilibrado.

#### Elección del Modelo:
Si bien el Soft Voting destaca en las métricas de entrenamiento, la evaluación definitiva realizada contra el set de prueba original (realidad del mercado) confirma que el Bagging sobre Random Forest es la solución más robusta y confiable. A diferencia de los modelos que tienden al sobreajuste en entornos ideales, el Bagging mantiene un mejor equilibrio y estabilidad ante el desbalance real de la base de datos. Por lo tanto, se recomienda su implementación final para optimizar las campañas de telemarketing, asegurando una captura eficiente de clientes sin comprometer la capacidad operativa por exceso de llamadas fallidas.

In [18]:
import pickle, os

# Guardamos el modelo Bagging y su umbral óptimo juntos
os.makedirs('../models', exist_ok=True)

with open('../models/bagging_rf_model.pkl', 'wb') as f:
    pickle.dump(bagging_rf, f)

# Guardamos el umbral para usarlo en producción
with open('../models/bagging_rf_threshold.pkl', 'wb') as f:
    pickle.dump(bagging_rf_threshold, f)

print("Modelo Bagging guardado: ../models/bagging_rf_model.pkl")
print(f"Umbral óptimo guardado: ../models/bagging_rf_threshold.pkl  (valor: {bagging_rf_threshold:.3f})")
print("Para predecir en producción:")
print("  probs = modelo.predict_proba(X_nuevo)[:, 1]")
print("  preds = (probs >= umbral).astype(int)")


Modelo Bagging guardado: ../models/bagging_rf_model.pkl
Umbral óptimo guardado: ../models/bagging_rf_threshold.pkl  (valor: 0.204)
Para predecir en producción:
  probs = modelo.predict_proba(X_nuevo)[:, 1]
  preds = (probs >= umbral).astype(int)
